### ISSUE TITLE
Add an after market thermometer in the cockpit.

### ISSUE THREAD

**Issue by wlbragg** [Juanvvc/c172p-detailed#626]
This was mentioned on a form post
http://forum.flightgear.org/viewtopic.php?f=17&t=28187

I think it would be a nice addition to have some kind of thermometer in the cockpit.

I never really even looked at the one on the vent to know for sure, but according to @wkitty42 it is the external temp.

**Comment by gilbertohasnofb** [comment]
Good idea. Which kind of model do you have in mind? Something digital like a small clock or something like this? https://s-media-cache-ak0.pinimg.com/236x/5a/c2/fd/5ac2fd4b564ee7ff5f7e16b52b257b08.jpg

I just would like to hear the opinion of @tigert or @Juanvvc about this, have you guys ever seen such thermometer in a cockpit?

As for the thermometer in the right vent, it's indeed external temperature.

**Comment by onox** [comment]
> As for the thermometer in the right vent, it's indeed external temperature.

That's correct, it's the external temperature.

**Comment by wkitty42** [comment]
On 12/14/2015 07:07 AM, Gilberto Agostinho wrote:
> Good idea. Which kind of model do you have in mind? Something digital like a
> small clock or something like this?

take the one out of the Piper PA24-150 Comanche... it is a stick-on there on the 
left window... black round disk with a LCD screen and a pushbutton to toggle it 
through C or F or Off...

stick it to the dash panel and tie it to the internal temperature calculations 
instead of the OAT as it is now...

**Comment by wkitty42** [comment]
On 12/14/2015 07:55 AM, onox wrote:
>     As for the thermometer in the right vent, it's indeed external temperature.
>
> That's correct, it's the external temperature.


especially since it says OAT right on it :wink:

**Comment by kjd-nc** [comment]
If one is concerned about the fact that such a thermometer is not a standard (or not common) piece of equipment, then would it be a big deal for it to be an optional feature of the console, defaulted to off?

As for digital vs a fluid thermometer, my first thought is, whichever is easier to read. Personally I wouldn't lose any sleep over whether there is a control to switch from C to F - I'm perfectly happy to set my preference in a menu or even in the set.xml file, and leave it there. 

If we treat it as a stick-on option, then it seems it would not be wired in to the console lights, therefore you'd need to use the flashlight to see it in the dark. Seems reasonable to me.

**Comment by gilbertohasnofb** [comment]
> If one is concerned about the fact that such a thermometer is not a standard (or not common) piece of equipment, then would it be a big deal for it to be an optional feature of the console, defaulted to off?

That's also my concern (hence my PM to you earlier today :wink: ). Maybe we can add a check box (default state = unchecked) in the aircraft menu which would then add the thermometer.

Probably a simple car dashboard thermometer would do the trick:

![](http://thumbs3.ebaystatic.com/d/l225/m/mALhqCaYLscR8AS5rWwG0XA.jpg)

According to the instructions of this specific little model:

> Display -19.9°C ~ 80°C.
>To press the first button is to switch the temperature on Celsius(°C) to Fahrenheit(°F)
>To press the second button for a little long time is to reset the temperature.

**Comment by wkitty42** [comment]
oops! that should be the "Piper PA24-250 Comanche" i referenced above... FGADDON/Aircraft/pa24-250... the stick-on is what they have as the OAT in that craft... looks like the code for it is all in Models/oat... it shouldn't be too hard to change "oat" to "iat" everywhere and then tie it with the existing IAT calculation code??

**Comment by gilbertohasnofb** [comment]
@wlbragg Can you post a screenshot of this thermometer you are recommending?

**Comment by wlbragg** [comment]
I don't have one in mind.
I did a search on the net for one that might be commonly used in aviation and all I found was external, that must be the main focus when it comes to temp. So I don't care what we use except that if it is not standard then it should be an option. 
Maybe they are not widely used because you can feel the temp with your senses. I though it might be useful in the simulation case because we can't feel the temp and need a reference.

**Comment by onox** [comment]
The temperature property for the cabin air is `/fdm/jsbsim/heat/cabin-air-temp-degc`.

**Comment by gilbertohasnofb** [comment]
@wlbragg I actually meant @wkitty42, my mistake. So @wkitty42 can you post a screenshot of the thermometer you recommended from that Comanche?

**Comment by wkitty42** [comment]
yeah... the craft is in fgaddon, though... but here is the screen shot...

![PA24-250 stick-on thermometer](https://www.dropbox.com/s/rkb3jlmu4bs4ius/fgfs-screen-099.png?dl=1)

push the white button to switch to F and off... it also turns off by itself after a few minutes just like a real one you can get at the drug, grocery or auto parts store...

there's an .ac file for it in the previously mentioned directory so it should be easy enough to make it rectangular if desired... my main thing is that a thermometer is available and should be easy enough to put in our craft and tie to the IAT stuffs rather than the OAT stuffs...

**Comment by wlbragg** [comment]
FWI: I don't think I would stick it to a window though, in RL it would be colder than the internal air temp, I would think.

**Comment by wkitty42** [comment]
that's why i said stick it on the dash... the above one is in the PA24-250 Comanche specifically for OAT as there's no such instrument in that craft :wink: 

the xml files that come with it appear to have positioning x,y,z coordinates in them... there's also a section or two in another xml file to include it as well as providing some sort of animation... at least, it is listed within animation tags but the only animation i see is pressing the white button and making the LCD characters turn on or off...

**Comment by wlbragg** [comment]
Sorry, wasn't reading everything.

**Comment by gilbertohasnofb** [comment]
@wkitty42 Thanks for the screenshot, it looks good to me. 

So gang, are we decided about adding it as an option in the aircraft dialog? If so, I volunteer to do that myself (and change the code for the cabin temp).

> FWI: I don't think I would stick it to a window though, in RL it would be colder than the internal air temp, I would think.

Where would you stick it then, to the panel itself? If so, what about here?:

![fgfs-screen-001](https://cloud.githubusercontent.com/assets/5700795/11824805/2503d85c-a37c-11e5-88ba-24fc81c219e7.png)

**Comment by onox** [comment]
Is the checkbutton really necessary? If it's visible as an instrument on that red spot, I don't mind it being always there.

**Comment by onox** [comment]
Can we reserve the empty spot on the left of it for a future carburetor temperature instrument?

**Comment by gilbertohasnofb** [comment]
> Is the checkbutton really necessary? If it's visible as an instrument on that red spot, I don't mind it being always there.

The thing is this isn't a standard instrument, hence why we have been talking about adding it as an option. In the end it isn't that much of a big deal.

> Can we reserve the empty spot on the left of it for a future carburetor temperature instrument?

Sure.

**Comment by wkitty42** [comment]
On 12/15/2015 04:34 PM, wlbragg wrote:
> Sorry, wasn't reading everything.

no problem... no need to apologize :wink:

**Comment by gilbertohasnofb** [comment]
Yeah, I know, we have tons of check buttons, but still... But anyway, I will just add the thermometer there and then we see. I improve its texture and now I am trying to adapt the code from the Comanche to our plane. As soon as I am done I will create a PR so you guys can check it out.

**Comment by gilbertohasnofb** [comment]
Hmm, I am having some difficulties with this one. The model is already in the right position, the new texture looks good. The Nasal code that controls the switch work fine, but I can't make the digital display work! I will create a PR for it, and I'd be very grateful if someone (@onox, wink, wink) could help me out here :smile:

**Comment by onox** [comment]
> > Can we reserve the empty spot on the left of it for a future carburetor temperature instrument?
>
> Sure.

@gilbertohasnofb I now remember that the amphibious already uses that spot :cry:

**Comment by gilbertohasnofb** [comment]
> I now remember that the amphibious already uses that spot :cry:

That's not a big problem, I can move it elsewhere. I will think about it for a second and propose something here.

**Comment by gilbertohasnofb** [comment]
So which one do you prefer?

![fgfs-screen-002](https://cloud.githubusercontent.com/assets/5700795/11829096/4fcf78ca-a39a-11e5-885f-a01215efaccc.png)

![fgfs-screen-001](https://cloud.githubusercontent.com/assets/5700795/11829101/554dd80a-a39a-11e5-9dbb-3ab874f8ed8d.png)

I think the top one is better because it's visible even if the yoke is on. The second option is awkward and in RL would be a nightmare for the pilot to look at.

**Comment by onox** [comment]
How about the position above the right yoke? (To the right of the ADF radio)

**Comment by gilbertohasnofb** [comment]
> How about the position above the right yoke? (To the right of the ADF radio)

There is a poster there in the OH-CTL livery added by @tigert, and I think it's a very nice touch that I wouldn't like to be deleted.

**Comment by onox** [comment]
Argh, we're running out of space :laughing: Hmm, I think neither positions are good. I'm not sure how readable the one on the left side is.

How about putting it above the right cockpit door? (A bit of an odd place though) Or between the trim and handheld mic?

**Comment by gilbertohasnofb** [comment]
> How about putting it above the right cockpit door? (A bit of an odd place though) Or between the trim and handheld mic?

I thought about those, but it looks awkward as hell... :sob:

**Comment by gilbertohasnofb** [comment]
What about putting it similarly to the top image above, but on the right wall instead of the left?

**Comment by onox** [comment]
That's a good compromise :+1: :smile:

**Comment by onox** [comment]
Can you make it a 3D physical object? (Not a flat 2D picture)

**Comment by gilbertohasnofb** [comment]
Okay, I will do it now, and then go to sleep :sleeping:

**Comment by gilbertohasnofb** [comment]
> Can you make it a 3D physical object? (Not a flat 2D picture)

What do you mean?

**Comment by gilbertohasnofb** [comment]
It is a 3D model.

**Comment by onox** [comment]
Okay, it looked like it was flat when you put in its original instrument position.

**Comment by gilbertohasnofb** [comment]
I mean, the model is really simple, it's just a shallow cylinder with a single button. Not even the "BAT" button is 3D. But I guess it's good enough for what we need, and people can improve it later.

**Comment by gilbertohasnofb** [comment]
Done and pushed, see: 
![fgfs-screen-001](https://cloud.githubusercontent.com/assets/5700795/11829828/c641d746-a39f-11e5-87bf-153462a5a2cd.png)

That's it for me for today, see you all tomorrow :smile:

**Comment by onox** [comment]
Sweet dreams @gilbertohasnofb :smile:

**Comment by gilbertohasnofb** [comment]
Thanks! :zzz:

**Comment by onox** [comment]
We have a branch `bug-626` that contains some work already done. The file `Models/Interior/Panel/Instruments/thermometer/thermometer.xml` contains stuff used by the old 2D panel. That needs to be deleted since it is useless.

Once the issue has been fixed, PR #630 can be reopened.

**Comment by gilbertohasnofb** [comment]
@onox @wlbragg I have just pushed new commits to branch `bug-626`. I have repositioned the thermometer (it was floating above the copilot's seat), made some improvements to its texture and added a new texture to the thermometer folder with all 10 digits as well as negative, dot and degree signs (which are needed for the temperature), as well as a F and C letters. I also tested that the function mode button was working, so the Nasal aspect is also done. 

All that is missing is to add the digit plaques to the ac model and create the logic. Here is how it will end up looking like:

![thermometer](https://user-images.githubusercontent.com/5700795/37244917-401de0b2-2488-11e8-9186-fd2b23e54a6e.png)
 
Anyone volunteers to take over from here? I am afraid I just don't know Blender enough to carry on from here.

**Comment by onox** [comment]
@gilbertohasnofb I think you need to split `window_base` into several objects. (And delete `sphere7` which doesn't seem to be visible). Then you can use the [textranslate](http://wiki.flightgear.org/Animations#Textranslate) animation for each of the digits.

**Comment by onox** [comment]
Why don't you give it a try? It's a separate model (not the main mesh) so you cannot really mess things up.

**Comment by onox** [comment]
You can select `window_base` and then do subdivide with 6 cuts, then select some edges and press x and choose "Disolve Eges". Then split the objects (select the 4 vertices of a digit, then press p and choose "Selection").

Maybe @wlbragg knows some better techniques.

**Comment by gilbertohasnofb** [comment]
@onox thanks for all the info, I will give this a shot.

**Comment by wlbragg** [comment]
>Maybe @wlbragg knows some better techniques.

Sounds about right considering I have been doing this exact step repeatedly over the last couple weeks, subdivide repeatedly and then clean up the mess.

I haven't looked at the model itself yet. @gilbertohasnofb if you get stuck let me know and I can help you with with it. OH, and there is an indoor/outdoor panel thermometer I hooked up in the j3cub you might be able to use for coding reference.

**Comment by gilbertohasnofb** [comment]
thanks, I will take a look at that!

**Comment by gilbertohasnofb** [comment]
@onox I believe this will involve not only creating the digit objects, but also UV-mapping. I am sorry but UV-mapping in Blender is beyond what I am capable of, and learning it will take quite a while.

If someone wants to create these objects for the digits, we need 33 of them:

* digits 0 through 9 in the first decimal place after the dot
* digits 0 through 9 in the units place
* digits 0 through 9 in the dozens place
* a minus sign which can be toggled
* Celcius and Farhenheit symbols

I can then implement the logic of when each is visible.

**Comment by onox** [comment]
You don't really need 10 objects for 1 digit. Just 1 rectangular object (a quad) and then you use a textranslate animation to shift the correct digit (from the texture) onto the object.

**Comment by gilbertohasnofb** [comment]
I see, thanks a lot. I will see if I manage this. But the UV-mapping in Blender kills me, I just have no earthly clue on how to do that...

**Comment by wlbragg** [comment]
@gilbertohasnofb Just looking at this and I need to ask a couple questions.
Why do we need to subdivide the display face? If your doing a text-translate you don't necessarily have to divide it do you? Can't we simply UV map the digits to where they need to go and text-translate from there when the display changes? Unless it is because you want a distinct line between the digits. If so, do you want these parts not only cut but slightly separated from each other. 
Or a completely different approach, why are we using textures at all for the display VS a canvas digital display? I think it would be less work all around to use canvas for the display. I did this for some digital displays in the AirCrane so I know how to do it. Is there any reason not to?

**Comment by gilbertohasnofb** [comment]
Possibly the reason to all those questions is that I am incompetent :wink: 

I was using our radios as reference to how digits work, e.g. the KX165, and the digits are separated objects which are then mapped separated using textranslate. I never used textranslate before so I wouldn't know if what you propose above is doable. As for using Canvas, sure we could do that but that's beyond my area. If someone would like to take over and make a canvas thermometer, then by all means go for it. I would be happy to help with textures if necessary.

**Comment by wlbragg** [comment]
>If someone would like to take over and make a canvas thermometer, then by all means go for it.

My thought on doing part of this in canvas is only for the digital display itself, it would be a lot less steps than trying to get the text-translation mappings, everything else would be the same.  @onox, any comments, suggestions.

**Comment by wlbragg** [comment]
@gilbertohasnofb 

Not trying to convince you here, just want to educate you as to how simple it can be to get the digital display in canvas VS texttranslate.

This is all that is required to update the temp display.

First the nasal control the display...
~~~
var digital_display =
  {

    new : func(designation, model_element, parameter, num_display, num_format)
      {
        var obj = {parents : [digital_display] };
        obj.designation = designation;
        obj.model_element = model_element;
        obj.parameter = parameter;
        obj.num_format = num_format;

        var dev_canvas= canvas.new({
          "name": designation,
          "size": [128,128], 
          "view": [128,128],                        
          "mipmapping": 0     
        });

        dev_canvas.addPlacement({"node": model_element});
        dev_canvas.setColorBackground(0,0,0,0);

        obj._canvas = dev_canvas;

        var root = dev_canvas.createGroup();

        obj.string = root.createChild("text")
        .setText(num_display)
        .setColor(.2,1,0.2)
        .setFontSize(30)
        .setScale(1,3)
        .setFont("DSEG/DSEG7/Classic-MINI/DSEG7ClassicMini-Bold.ttf")
        .setAlignment("center-bottom")
        .setTranslation(65, 105);

        obj.string.enableUpdate();

        return obj;
      },

  display : func()
    {
      var string =  sprintf(me.num_format, getprop("/fdm/jsbsim/heat/"~me.parameter));
      me.string.updateText(string);
    },

  update: func()
    {
      me.display();
      settimer (func me.update(), 0.1);
    },

};

var temp_display = digital_display.new("Temperature Display", "digital-display-glass", "cabin-air-temp-degc", "0.00", "%1.2f");

temp_display.update();
~~~

and in temperature.xml same object we were going to texttranslate on (more or less).
~~~
  <animation>
    !-- was window_base object -->
    <object-name>digital-display-glass</object-name>
    <type>material</type>
    <condition>
      <and>
        <not>
          <property>/sim/rendering/rembrandt/enabled</property>
        </not>
      </and>
    </condition>
    <emission>
      <red>1.0</red>
      <green>1.0</green>
      <blue>1.0</blue>
      <factor-prop>/systems/electrical/outputs/digital-display</factor-prop>
    </emission>
  </animation>
~~~

I think everything else is going to be pretty mush the same.

**Comment by wlbragg** [comment]
I do still want to get @onox's opinion.

**Comment by onox** [comment]
@wlbragg Let's do canvas!

**Comment by onox** [comment]
But please put unindent the braces, you're using Thorsten indentation style :stuck_out_tongue:

**Comment by onox** [comment]
And use `maketimer` instead of `settimer`.

**Comment by gilbertohasnofb** [comment]
Just so this doesn't get lost, my work on these textures was pushed to `bug-626`, in case anyone feels like continuing from there.

**Comment by legoboyvdlp** [comment]
What needs done to finish it?

**Comment by gilbertohasnofb** [comment]
Basically the implementation of the code. The textures are pushed to that branch, but as the others have been talking in the recent messages above using Canvas would be the best option. Unfortunately I don't know anything about that and I have already too much on my plate with all these 3D and texture issues that we are currently dealing with. If you feel like having a go at this, go for it!

**Comment by legoboyvdlp** [comment]
If you give me a couple of months, I'll try it (exam season currently!)

**Comment by gilbertohasnofb** [comment]
Nobody has touched this in a while so there's a chance it might still be opened. When you are free to give this a go just make sure to ask if no one else started working on it again, ok?

**Comment by legoboyvdlp** [comment]
Wilco!

**Comment by wlbragg** [comment]
>If you give me a couple of months, I'll try it (exam season currently!)

We are going to do this with canvas an I'll probably include this in the next round of interior work. I was just waiting for the opportunity.

**Comment by gilbertohasnofb** [comment]
@wlbragg Sounds good to me

**Comment by gilbertohasnofb** [comment]
@legoboyvdlp are you still interested in creating a Canvas thermometer?

**Comment by legoboyvdlp** [comment]
Not able to at the moment I'm afraid.

**Comment by gilbertohasnofb** [comment]
No problems, let's leave this open and see if you or someone else picks it up one day, no rush at all.
